File: Judge/TestCase.cpp
#include "TestCase.h"

#include <fstream>
#include <iostream>
#include <sstream>
#include <ios>
#include <regex>


namespace sjtu {

TestCase parse(const std::string &dir) {
    std::ifstream fin(dir);
    if (!fin.is_open())
        throw std::ifstream::failure(std::string("file does not exists"));
    std::string buffer(static_cast<std::stringstream const &>(std::stringstream() << fin.rdbuf()).str());

    TestCase result;
    result.src = readSource(buffer);

    CIter beg, end;

    std::tie(beg, end) = findBlock(buffer, "comment");
    result.comment = std::string(beg, end);

    std::tie(beg, end) = findBlock(buffer, "input");
    result.input = std::string(beg, end);

    std::tie(beg, end) = findBlock(buffer, "output");
    result.output = std::string(beg, end);

    std::tie(beg, end) = findBlock(buffer, "assert");
    if (beg != end) {
        std::string rawAssertion(beg, end);
        if (rawAssertion == "success_compile")
            result.assertion = AssertionType::SuccessCompile;
        else if (rawAssertion == "failure_compile")
            result.assertion = AssertionType::FailureCompile;
        else if (rawAssertion == "exitcode")
            result.assertion = AssertionType::Exitcode;
        else if (rawAssertion == "runtime_error")
            result.assertion = AssertionType::RuntimeError;
        else if (rawAssertion == "output")
            result.assertion = AssertionType::Output;
        else
            throw InvalidTestCase();
    }

    std::tie(beg, end) = findBlock(buffer, "timeout");
    if (beg != end) {
        std::string rawTimeout(beg, end);
        auto dTimeout = std::stod(rawTimeout, nullptr);
        result.timeout = (std::uint64_t) (dTimeout * 1000);
    }
    
    std::tie(beg, end) = findBlock(buffer, "exitcode");
    if (beg != end) {
        std::string rawExitcode(beg, end);
        result.exitcode = std::stoi(rawExitcode, nullptr);
    }

    std::tie(beg, end) = findBlock(buffer, "phase");
    std::string rawPhase(beg, end);
    if (rawPhase == "semantic pretest")
        result.phase = TestCasePhase::SemanticPretest;
    else if (rawPhase == "semantic extended")
        result.phase = TestCasePhase::SemanticExtended;
    else if (rawPhase == "codegen pretest")
        result.phase = TestCasePhase::CodegenPretest;
    else if (rawPhase == "codegen extended")
        result.phase = TestCasePhase::CodegenExtended;
    else if (rawPhase == "optim pretest")
        result.phase = TestCasePhase::OptimPretest;
    else if (rawPhase == "optim extended")
        result.phase = TestCasePhase::OptimExtended;
    else
        throw InvalidTestCase();

    return result;
}

std::string readSource(const std::string &buffer) {
    std::smatch m;
    if (!std::regex_search(buffer, m, std::regex("/\\*!! metadata:")))
        throw InvalidTestCase();
    return {buffer.begin(), m[0].first};
}

std::pair<CIter, CIter> findBlock(const std::string &buffer, const std::string &name) {
    std::smatch m;
    if (!std::regex_search(buffer, m,
                           std::regex("=== " + name + " ===\n(.|\n)*?===")))
        return {buffer.end(), buffer.end()};
    auto beg = m[0].first;
    auto end = m[0].second;

    std::string tmp(beg, end);
    std::cout << tmp << std::endl;

    while (*beg != '\n')
        ++beg;
    ++beg;
    while (*end != '\n')
        --end;
    return {beg, end};
}


} // namespace sjtu

File: compiler_oj/testcase.py
import os


class TestCase:
    def __init__(self, raw, filename="unknown", t=1.0):
        self.__raw = raw
        self.filename = filename
        self.src = self.__read_source()
        self.comment = self.__find_block("comment")
        self.input = self.__find_block("input")
        self.output = self.__format_output(self.__find_block("output"))
        self.assertion = self.__find_block("assert")
        self.timeout = self.__find_block("timeout")
        if self.timeout != "":
            self.timeout = float(self.timeout) 
        else:
            self.timeout = 10.0
        self.timeout /= t # Continue for t seconds 续一秒!!
        self.exitcode = self.__find_block("exitcode")
        self.exitcode = int(self.exitcode) if self.exitcode != "" else None
        self.phase = self.__find_block("phase")

    @staticmethod
    def __format_output(raw):
        return '\n'.join(list(map(lambda x: x.strip(), raw.split('\n'))))

    def __read_source(self):
        end = self.__raw.find("/*!! metadata:")
        return self.__raw[0:end]

    def __find_block(self, name):
        title = '=== ' + name + ' ==='
        beg = self.__raw.find(title)
        if beg == -1:
            return ''
        beg += len(title)
        end = self.__raw.find("===", beg)
        if end == -1:
            end = self.__raw.find("!!*", beg)
        if end == -1:
            return ""
        return self.__raw[beg:end].strip()


def read_testcases(dirs, t):
    testcases = []

    for dir in dirs:
        names = os.listdir(dir)
        print(dir + " : " +  str(len(names)))
        for name in names:
            __, extension = os.path.splitext(name)

            if extension != ".txt" and extension != ".mx":
                # print(extension)
                continue

            with open(os.path.join(dir, name)) as f:
                raw = f.read()
                testcases.append(TestCase(raw, name, t))

    print("testcases at all: " + str(len(testcases)))
    # print([i.phase for i in testcases])
    return testcases
